Partitioning before Merging

samanvp opened the pull request:
In order to be able to process large datasets (#78), we need to partition
input into smaller patches so that Merging would be possible to run on a
ordinary VM. Here we add the ability to partition variants based on
their reference_name (CHROM).

This change enables VT to process Platinum_100 (with a caviate of some
changes needed for writing to BigQuery table).

samanvp commented on gcp_variant_transforms/vcf_to_bq.py:
```diff
@@ -248,8 +252,15 @@ def run(argv=None):
+      merged = [(partitions[i] | 'MergeVariants'+str(i) >>
+                 merge_variants.MergeVariants(variant_merger))
+                for i in xrange(_DEFAULT_PARTITION_NO)]
+      variants = merged | 'FlattenPartitions' >> beam.Flatten()
```
Note this implementation does not make any guarantee that each chromosome will be assigned to a separate partition. If reference names do not follow the standard format ('^(c|ch|chr)?([0-9][0-9]?)$') we hash them and randomly assign them to a bucket (without checking for collisions). Even when the reference names have standard format we assign chromosome 1-22 to partitions [0, 21] and the remaining ones (chrX, chrY, chrM, or any variants with unknown_... reference_name) are assigned randomly to other partition indices > 21.

For the use case you mentioned, we will be using the partition_config.yaml file (which is being implemented right now). We will offer a default config file so that users can partition their output table by just setting a flag to the default config file.

bashir2 replied:
Thanks Saman for the clarifications. What I am trying to get is to avoid parsing a config file for chromosome based partitioning. Have you thought about generating list of chromosome strings (possibly with count) and feeding it back to the pipeline using side-inputs? I have added both topics related to partitioning to the weekly meeting agenda, we can discuss further there.

arostamianfar replied:
Bashir, what is the drawback of using a config file? We'd need to implement that logic anyway (and it's not that complicated), but does provide a robust way of partitioning. Note that even the 'hg19 reference' genome has a log of random chromosome names: http://hgdownload.cse.ucsc.edu/goldenpath/hg19/chromosomes/

arostamianfar commented on gcp_variant_transforms/transforms/partition_variants.py:
```diff
@@ -0,0 +1,73 @@
+  def __init__(self):
+    """Initializes the transform, mainly a dic of chrom->index for quick look up."""
+
+    self._by_chr = collections.defaultdict(int)
```
It doesnt' look like you're actually using the defaultdict functionality, so why not just use a simple dict?

arostamianfar commented on gcp_variant_transforms/transforms/partition_variants_test.py:
```diff
@@ -0,0 +1,114 @@
+    variants.append(vcfio.Variant(reference_name='ch2'))
+    variants.append(vcfio.Variant(reference_name='CH2'))
+    # Partition 2
+    variants.append(vcfio.Variant(reference_name='c3'))
```
It may actually be better to use some more realistic examples here as well.

arostamianfar commented on gcp_variant_transforms/transforms/partition_variants_test.py:
```diff
@@ -0,0 +1,114 @@
+    variants.append(vcfio.Variant(reference_name='chr22    '))
+    return variants, variants[0:2], variants[2:4], variants[4:6], variants[6:9], variants[9:11], variants[11:13]
+
+  def _get_invalid_variants(self):
```
consider rewording 'valid' and 'invalid'  (perhaps 'nonstandard' and 'standard'?)

arostamianfar commented on gcp_variant_transforms/transforms/partition_variants_test.py:
```diff
@@ -0,0 +1,114 @@
+    # they will all end up in the default partition.
+    return variants, variants
+
+  def test_partition_variants(self):
```
I think the test can be written in a more readable manner. Let's chat offline.

arostamianfar commented on setup.py:
```diff
@@ -25,6 +25,7 @@
     'google-api-python-client>=1.6',
     'intervaltree>=2.1.0,<2.2.0',
     'pyvcf<0.7.0',
+    'mmh3>=2.5.1',
```
Given the random breakages that we've had in the past due to pip packages getting updated, could we pin this particular version instead? Or have a lower limit of versions? I suspect it's not used by any of our dependencies.

samanvp replied:
Thanks for clarifying.

arostamianfar commented on gcp_variant_transforms/vcf_to_bq.py:
```diff
@@ -74,7 +75,10 @@
+# The first 22 partitions [0, 21] are reserved for standard reference_name
+# formated as RegExp '^(c|ch|chr)?([0-9][0-9]?)$'.
+# Every other identifers will be matched to next available partitions [22, 26].
+_DEFAULT_PARTITION_NO = 22 + 5
```
I see. Thanks for the extra context. However, I have been thinking about this and I'm a bit uneasy about having this random number in our main module, and later the PartitionConfig class added here as well. I think a better approach could be:
- Create a new libs.variant_partition.VariantPartition class that inherits from object (i.e. simple lib class).
- Define this constant in that module.
- Implement a `get_num_partitions()` method that, for now, just returns that constant. Later on, it can parse the config file and return the appropriate value.
- Implement a `partition(variant)` method inside that lib and move all of the partitioning logic there.
- Change PartitionVariants to get the VariantPartition lib as an arg. The partition_for method simply does `return variant_partition.partition(variant)`. You may actually choose to remove the ParitionVariants class altogether since it's just a dummy class now and can call VariantPartition.partition directly (you'd have a dummy num_partitions arg, but that's ok).

Finally, the code in main becomes something like:
```
variant_partition_lib = VariantPartition() # this can later take in config_file_path
partitions = variants | 'PartitionVariants' >> beam.Partition(variant_parition_lib.partition, variant_parition_lib.get_num_partitions())
merged = [... for i in xrange(variant_parition_lib.get_num_partitions())]
...
```

The benefit of this approach is that all of the partitioning logic lives in one place and we can add/remove logic about handling config files without affecting the main module.
WDYT? I'm very open to other ideas as well as long as it removes the constant from main :)

arostamianfar commented on gcp_variant_transforms/vcf_to_bq.py:
```diff
@@ -248,8 +252,15 @@ def run(argv=None):
-    variants |= (
-        'MergeVariants' >> merge_variants.MergeVariants(variant_merger))
+    if known_args.optimize_for_large_inputs:
+      partitions = variants | 'PartitionVariants' >> beam.Partition(
```
The main point is to increase readability as it gives a "name" and "context" for what this block of code is doing, but feel free to leave as is for now since it's pretty small :)

arostamianfar replied:
Sounds good!

samanvp commented on gcp_variant_transforms/transforms/partition_variants.py:
```diff
@@ -0,0 +1,82 @@
+      ValueError: If partition_no <= _RESERVED_PARTITIONS
+    """
+
+    chrom = (variant.reference_name).strip().lower()
```
I don't think so, I thought it will make it more readable, now that I look at it I don't see how this is more readable :)

samanvp commented:
Following up on your comment of refactoring the partitioning logic from the actual Beam.partition_for() I added a new class to this PR. I was planning to add this new class in a follow up PR which implements reading the partition_config.yaml file as well as applying to to the pipeline.
To avoid redundant work, I added that class to this PR along with all of its unit tests. However, the logic of actually utilizing a config_file (through a flag) and exporting the final output to multiple BQ table is not added to this PR and will be send out in a future one.

arostamianfar commented on gcp_variant_transforms/transforms/partition_variants_test.py:
```diff
@@ -0,0 +1,121 @@
+    # Partition 21
+    variants.append(vcfio.Variant(reference_name='chr22', start=0))
+    variants.append(vcfio.Variant(reference_name='chr22    ', start=0))
+    return variants, variants[0:2], variants[2:4], variants[4:6],\
```
Instead of this, I think it's much more readable to return a dict where the key is the partition number (int) and the value is a list of variants that should map to that partition. You can just return that dict both here and in non-standard case.
In your test, you would essentially pass the dict values as inputs to the pipeline, and then loop through [0, get_num_partitions) and if the key is in the dict, make sure that it matches the value, if not make sure it's empty. WDYT?
P.S. Is checking the 'label' important? If so, I suppose you can have a tuple of partition number and label as the key to your dict.

arostamianfar commented on gcp_variant_transforms/libs/variant_partition.py:
```diff
@@ -0,0 +1,91 @@
+from __future__ import absolute_import
+
+import re
+from mmh3 import hash  # pylint: disable=no-name-in-module,redefined-builtin
```
huh intersting!

arostamianfar commented on gcp_variant_transforms/libs/variant_partition.py:
```diff
@@ -0,0 +1,92 @@
+
+  def get_partition(self, reference_name, pos=0):
+    reference_name = reference_name.strip().lower()
+    if ((not reference_name) or (pos < 0)):
```
nit: no need for surrounding brackets (styleguide is picky about avoiding unnecessary surrounding brackets :) )
`if not reference or pos < 0:` should work.

arostamianfar commented on gcp_variant_transforms/transforms/partition_variants.py:
```diff
@@ -0,0 +1,33 @@
+    self._partition = partition
+
+  def partition_for(self, variant, num_partitions):
+    # type: (Variant, int) -> int
```
nit: s/Variant/vcfio.Variant

arostamianfar commented on gcp_variant_transforms/transforms/partition_variants_test.py:
```diff
@@ -0,0 +1,130 @@
+    expected_partitions[23] = []
+    return variants, expected_partitions
+
+  def test_partition_variants(self):
```
This is still a bit difficult to read; consider the following approach using some python magic :D
```
expected_partitions = {}
expected_partitions[0] = [vcfio.Variant(reference_name='chr1', start=0),
                          vcfio.Variant(reference_name='CHR1', start=0)]
...etc
# do not add empty partitions as it can be error prone (e.g. what if you change the default #?)
return expected_partitions
```
In your test then:
```
variant_partitions = _get_standard_variant_partitions()
variant_partitions.update(_get_non_standard_variant_partitions())
variants = [variant for variant_list in variant_partitions for variant in variant_list]  # essentially flattens the lists
...
pipeline | Create(variants) 
...
for i in xrange(partitioner.get_num_partitions()):
  assert_that(partitions[i], equal_to(variant_partitions.get(i, [])), label=str(i))

```

samanvp replied:
done.

arostamianfar commented on gcp_variant_transforms/transforms/partition_variants_test.py:
```diff
@@ -0,0 +1,146 @@
+    return expected_partitions
+
+  def test_partition_variants(self):
+    standard_variants = self._get_standard_variants()
```
you dont actually need these extra variants; can just get them from `expected_partitions.values()`
you just need to 'flatten' them, which you can do with that for loop i had mentioned previously :)
